src/psyclone/tests/psyir/transformations/intrinsics/dotproduct2code_trans_test.py

```diff
@@ -0,0 +1,451 @@
+    psyir = fortran_reader.psyir_from_source(code)
+    dot_product = psyir.walk(BinaryOperation)[0]
+    assert dot_product.operator == BinaryOperation.Operator.DOT_PRODUCT
+    lower, upper, step = _get_array_bound(
+        dot_product.children[0], dot_product.children[1], fortran_writer)
+    assert fortran_writer(lower) == 'LBOUND(v1, 1)'
+    assert fortran_writer(upper) == 'UBOUND(v1, 1)'
+    assert step.value == '1'
+
+
+# DotProduct2CodeTrans class init method
+
+def test_initialise():
+    '''Check that the class DotProduct2CodeTrans behaves as expected when
+    an instance of the class is created. Note, this also tests that
+    the parent constructor is also called as this sets the name and
+    value of __str__.
+
+    '''
+    trans = DotProduct2CodeTrans()
+    assert trans._operator_name == "DOTPRODUCT"
+    assert (str(trans) == "Convert the PSyIR DOTPRODUCT intrinsic to "
+            "equivalent PSyIR code.")
+    assert trans.name == "Dotproduct2CodeTrans"
+
+
+# DotProduct2CodeTrans class validate method
+
+def test_validate_super():
+    '''Test that the DotProduct2CodeTrans validate method calls the validate
+    method of the parent class.
+
+    '''
+    trans = DotProduct2CodeTrans()
+    with pytest.raises(TransformationError) as info:
+        trans.validate(None)
+    assert ("The supplied node argument is not a DOTPRODUCT operator, found "
+            "'NoneType'." in str(info.value))
+
+
+def test_validate_references(fortran_reader):
+    '''Test that the DotProduct2CodeTrans validate method produces the
+    expected exception when at least one of the arguments is not an
+    array. In this case it is a matmul intrinsic.
+
+    '''
+    code = (
+        f"subroutine dot_product_test(v1,v2,a3)\n"
+        f"real,intent(in) :: v1(:), v2(:), a3(:,:)\n"
+        f"real :: result\n"
+        f"result = dot_product(matmul(a3,v1),v2)\n"
+        f"end subroutine\n")
+    expected = (
+        "The DotProduct2CodeTrans transformation only supports the "
+        "transformation of a dotproduct intrinsic if its arguments "
+        "are arrays, but found MATMUL(a3, v1) in "
+        "DOT_PRODUCT(MATMUL(a3, v1), v2)")
+    check_validate(code, expected, fortran_reader)
+
+
+def test_validate_1d_array(fortran_reader):
+    '''Test that the DotProduct2CodeTrans validate method produces the
+    expected exception when at least one of the arguments does not use
+    array slice notation but is not known to be a 1D array.
+
+    '''
+    code = (
+        f"subroutine dot_product_test(a1,a2)\n"
+        f"real,intent(in) :: a1(:,:), a2(:,:)\n"
+        f"real :: result\n"
+        f"result = dot_product(a1,a2)\n"
+        f"end subroutine\n")
+    expected = (
+        "The DotProduct2CodeTrans transformation only supports the "
+        "transformation of a dotproduct intrinsic with an argument not "
+        "containing an array slice if the argument is a 1D array, but "
+        "found a1 in DOT_PRODUCT(a1, a2).")
```

Comment:
"found 'a1' which has shape xxxx in DOT_PRODUCT..."

Reply:
Better, thanks.